Geometry module, link and board mapping updates

snwebb opened the pull request:
#### PR description:

Adding new functionality to obtain the mapping between modules, lpgbts, and stage 1 and 2 FPGAs.
These are in the form of a `.json' file, which is then read-in and manipulated (json file only an example skeleton at present)

- Possible improvements:
  - better naming of functions?
  - better integration with similar existing functions?
  - return module detid instead of u,v,layer (although not all information is available, e.g. z-side)

#### PR validation:

Tested that the code compiles and each function works in isolation.

snwebb committed:
start by copying existing file

snwebb committed:
First working version of input json file for the mapping

snwebb committed:
adding all possible combinations of functions

snwebb committed:
code format

snwebb committed:
Add mapping instead of reading directly from json, and update .json format

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,905 @@
+
+  unsigned layerWithOffset(unsigned) const;
+
+  std::vector<unsigned> getStage1FpgasFromStage2Fpga(const unsigned stage2_id) const;
+  std::vector<unsigned> getLbgbtsFromStage1Fpga(const unsigned stage1_id) const;
+  std::vector<unsigned> getStage2FpgasFromStage1Fpga(const unsigned stage1_id) const;
```
For the returned type, it would be better to stick to `geom_set` or `geom_ordered_set`, as used in other mapping interface functions (e.g. `getTriggerCellsFromModule()`)

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,905 @@
+
+  unsigned layerWithOffset(unsigned) const;
+
+  std::vector<unsigned> getStage1FpgasFromStage2Fpga(const unsigned stage2_id) const;
+  std::vector<unsigned> getLbgbtsFromStage1Fpga(const unsigned stage1_id) const;
+  std::vector<unsigned> getStage2FpgasFromStage1Fpga(const unsigned stage1_id) const;
+  unsigned getStage1FpgaFromLpgbt(const unsigned lpgbt_id) const;
+  std::vector<std::tuple<bool, int, int, unsigned> > getModulesFromLpgbt(const unsigned lpgbt_id) const;
+  std::vector<unsigned> getLpgbtsFromModule(const bool isSilicon, const int module_u, const int module_v, const unsigned layer) const;
```
Module ids (integer) should be defined instead of accessing the mappings with tuples. This is to reduce as much as possible CPU time. See also my other comments below on the access functions.

jbsauvan replied:
All these access functions should  be public. They should also be defined in the base geometry class since this is the external interface users have access to.

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,905 @@
+
+std::vector<unsigned> HGCalTriggerGeometryV9Imp3::getStage1FpgasFromStage2Fpga(const unsigned stage2_id) const {
+
+  std::vector<unsigned> stage1_ids = mapping_config_["Stage2"][stage2_id]["Stage1"];
```
The mapping information should be stored into `unordered_map` or `unordered_multimap` for fast access. The translation from the json format to maps should be done only in the initialization.
Geometry accesses are among the most CPU consuming in the HGC TPG, so these mapping accesses should be as fast as possible. Unordered maps with integer keys should be used because of their constant access complexity.

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,905 @@
+
+std::vector<unsigned> HGCalTriggerGeometryV9Imp3::getLbgbtsFromStage1Fpga(const unsigned stage1_id) const {
+
+  std::vector<unsigned> lpgbt_ids = mapping_config_["Stage1"][stage1_id]["lpgbts"];
```
Same as above

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,905 @@
+
+std::vector<unsigned> HGCalTriggerGeometryV9Imp3::getStage2FpgasFromStage1Fpga(const unsigned stage1_id) const {
+
+  std::vector<unsigned> stage2_ids = mapping_config_["Stage1"][stage1_id]["Stage2"];
```
Same as above

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,905 @@
+
+unsigned HGCalTriggerGeometryV9Imp3::getStage1FpgaFromLpgbt(const unsigned lpgbt_id) const {
+
+  unsigned stage1_id = mapping_config_["lpgbt"][lpgbt_id]["Stage1"];
```
Same as above

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,905 @@
+  std::vector<std::tuple<bool, int, int, unsigned > > modules;
+
+  for (auto module : mapping_config_["lpgbt"][lpgbt_id]["Modules"] ){
+    modules.emplace_back(std::make_tuple(module["isSilicon"],module["u"],module["v"],module["layer"]));
```
Same as above

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,905 @@
+std::vector<unsigned> HGCalTriggerGeometryV9Imp3::getLpgbtsFromModule(const bool isSilicon, const int module_u, const int module_v, const unsigned layer) const {
+
+  std::vector<unsigned> lpgbts_ids;
+  for (auto module : mapping_config_["Module"]) {
+    if (module["isSilicon"]==isSilicon 
+	&&module["u"]==module_u
+	&&module["v"]==module_v
+	&&module["layer"]==layer){
+      for (auto lpgbt : module["lpgbts"]) {
+	lpgbts_ids.emplace_back(lpgbt["id"]);
+      }
+    }
+  }
```
Same as above

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,905 @@
+void HGCalTriggerGeometryV9Imp3::fillMaps() {
+  // read module mapping file
+  std::ifstream l1tModulesMappingStream(l1tModulesMapping_.fullPath());
+  if (!l1tModulesMappingStream.is_open()) {
+    throw cms::Exception("MissingDataFile") << "Cannot open HGCalTriggerGeometry L1TModulesMapping file\n";
+  }
+
+  short waferu = 0;
+  short waferv = 0;
+  short module = 0;
+  short layer = 0;
+  for (; l1tModulesMappingStream >> layer >> waferu >> waferv >> module;) {
+    wafer_to_module_.emplace(packLayerWaferId(layer, waferu, waferv), module);
+    module_to_wafers_.emplace(packLayerModuleId(layer, module), packWaferId(waferu, waferv));
+  }
+  if (!l1tModulesMappingStream.eof()) {
+    throw cms::Exception("BadGeometryFile")
+        << "Error reading L1TModulesMapping '" << layer << " " << waferu << " " << waferv << " " << module << "' \n";
+  }
+  l1tModulesMappingStream.close();
+  // read links mapping file
+  std::ifstream l1tLinksMappingStream(l1tLinksMapping_.fullPath());
+  if (!l1tLinksMappingStream.is_open()) {
+    throw cms::Exception("MissingDataFile") << "Cannot open HGCalTriggerGeometry L1TLinksMapping file\n";
+  }
+  short links = 0;
+  const short max_modules_60deg_sector = 127;
+  for (; l1tLinksMappingStream >> layer >> module >> links;) {
+    if (module_to_wafers_.find(packLayerModuleId(layer, module)) == module_to_wafers_.end()) {
+      links = 0;
+    }
+    if (module > max_modules_60deg_sector)
+      sector0_mask_ = 0xff;  // Use 8 bits to encode module number in 120deg sector
+    links_per_module_.emplace(packLayerModuleId(layer, module), links);
+  }
+  if (!l1tLinksMappingStream.eof()) {
+    throw cms::Exception("BadGeometryFile")
+        << "Error reading L1TLinksMapping '" << layer << " " << module << " " << links << "' \n";
+  }
+  l1tLinksMappingStream.close();
```
All these mapping filling should be replaced. All these information should in principle be retrieved from the json file.

snwebb committed:
make functions public

snwebb committed:
format

snwebb commented:
Thanks @jbsauvan for the comments, which I have now implemented.

I have a couple of points which I probably need to address, and you might have an opinion:

-  I have left in the correspondence between "modules" and "wafers" and added a `module_id` to the module entries in the json file, which would be the same as the one in the existing input file. 
- The user may wish to access the modules via `waferu` and `waferv`, so I could add an additional function if needed (and is there a corresponding id for the scintillator?)

jbsauvan commented:
I don't think we need a correspondence between modules and wafers in the json file since there is now a one to one equivalence. What is needed is:
-  using a detid for modules (and other objects) in order to access them globally in the detector. Modules (and other objects) in the whole detector should have unique ids. Since modules follows the same structure as wafers (or trigger cells without the TC u and v), a `HGCalTriggerDetId` could be used. But we could also think to a new id scheme, based on a sector structure, which may be more practical, and may be used not only for modules in the silicon section, but also in the scintillator section, as well as for links and boards if possible.
Currently, `HGCScintillatorDetId` are used for scintillator modules (https://github.com/PFCal-dev/cmssw/blob/a4ef51dc81738c074eabdfc9f0f88ec01a227542/L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc#L247), and `HGCalDetId` for silicon modules.
- having an internal reduced "id" to store efficiently mapping information, as it is done e.g. here (https://github.com/PFCal-dev/cmssw/blob/a4ef51dc81738c074eabdfc9f0f88ec01a227542/L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp2.cc#L274). For instance we need a mapping between lpgbt and modules only in one sector, the mapping in the other sectors can be deduced from that and don't need to be stored.

> The user may wish to access the modules via waferu and waferv, so I could add an additional function if needed (and is there a corresponding id for the scintillator?)

Passing a single detid is more generic as it doesn't assume a given id scheme. This is required in order to have a base/interface class that can be used for different id schemes. The specificity of each id scheme is handled in each geometry concrete class. Having a function taking u,v as parameters couldn't be used externally because the other geometry classes don't follow this scheme.

snwebb committed:
adding function to all geometry files to avoid crash

snwebb committed:
adding new HGCalModuleDetId class

snwebb committed:
update get trigger cells from module function

snwebb committed:
add all rotation functions for silicon and scintillator

snwebb committed:
fix trigger cell to modules functions

snwebb committed:
remove module id from json

snwebb committed:
update json readers

snwebb commented:
Hi @jbsauvan , Thanks for the comments and discussion.
I've updated the code taking these into account. I still think we will need another iteration, as there are some areas I am not really sure if we need, or how best to implement something (particularly for the HFNose).

snwebb commented:
Closing - continued in #466
